source_control/gitlab_project.py

```diff
@@ -0,0 +1,389 @@
+                import_url="http://git.example.com/example/lab.git"
+                state=present
+'''
+
+try:
+    import gitlab
+    HAS_GITLAB_PACKAGE = True
+except:
+    HAS_GITLAB_PACKAGE = False
+
+
+class GitLabProject(object):
+    def __init__(self, module, git):
+        self._module = module
+        self._gitlab = git
+
+    def createOrUpdateProject(self, project_exists, group_name, import_url, arguments):
+        is_user = False
+        group_id = self.getGroupId(group_name)
+        if group_id == '':
+            group_id = self.getUserId(group_name)
+            is_user = True
+
+        if project_exists:
+            # Edit project
+            return self.updateProject(group_name, arguments)
+        else:
+            # Create project
+            if self._module.check_mode:
+                self._module.exit_json(changed=True)
+            return self.createProject(is_user, group_id, import_url, arguments)
+
+    def createProject(self, is_user, user_id, import_url, arguments):
+        if is_user:
+            return self._gitlab.createprojectuser(user_id=user_id, **arguments)
+        else:
+            group_id = user_id
+            return self._gitlab.createproject(namespace_id=group_id, import_url=import_url, **arguments)
+
+    def deleteProject(self, group_name, project_name):
+        if self.existsGroup(group_name):
+            project_owner = group_name
+        else:
+            project_owner = self._gitlab.currentuser()['name']
+
+        search_results = self._gitlab.searchproject(search=project_name)
+        for result in search_results:
+            owner = result['namespace']['name']
+            if owner == project_owner:
+                return self._gitlab.deleteproject(result['id'])
+
+    def existsProject(self, group_name, project_name):
+        if self.existsGroup(group_name):
+            project_owner = group_name
+        else:
+            project_owner = self._gitlab.currentuser()['name']
+
+        search_results = self._gitlab.searchproject(search=project_name)
+        for result in search_results:
+            owner = result['namespace']['name']
+            if owner == project_owner:
+                return True
+        return False
+
+    def existsGroup(self, group_name):
+        if group_name is not None:
+            # Find the group, if group not exists we try for user
+            for group in self._gitlab.getall(self._gitlab.getgroups):
+                if group['name'] == group_name:
+                    return True
+
+            user_name = group_name
+            user_id = self._gitlab.getusers(search=user_name)['id']
+            if user_id != '':
+                return True
+        return False
+
+    def getGroupId(self, group_name):
+        if group_name is not None:
+            # Find the group, if group not exists we try for user
+            for group in self._gitlab.getall(self._gitlab.getgroups):
+                if group['name'] == group_name:
```

Comment:
are group names unique and case insensitive? I often need to add .lower() to both values.